File: saas-platform/scripts/db-manager.js
#!/usr/bin/env node

/**
 * Unified Database Management Tool
 * This is the ONLY script you need for database operations
 */

const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');
require('dotenv').config();

// Colors for terminal output
const colors = {
  reset: '\x1b[0m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
};

// Check environment
const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_KEY;
const SUPABASE_ANON_KEY = process.env.SUPABASE_ANON_KEY;
const SUPABASE_DB_PASSWORD = process.env.SUPABASE_DB_PASSWORD;

if (!SUPABASE_URL || !SUPABASE_SERVICE_KEY) {
  console.error(`${colors.red}❌ Missing required environment variables:${colors.reset}`);
  console.error('   SUPABASE_URL and SUPABASE_SERVICE_KEY must be set in .env');
  process.exit(1);
}

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY, {
  auth: {
    autoRefreshToken: false,
    persistSession: false
  }
});

// Command handlers
const commands = {
  /**
   * Show available migrations and instructions
   */
  async status() {
    console.log(`${colors.cyan}📊 Database Status${colors.reset}`);
    console.log('==================\n');

    // Check what tables exist
    const tables = await checkTables();
    const expectedTables = ['accounts', 'subscriptions', 'instances', 'usage_metrics', 'webhook_events', 'audit_logs'];

    console.log('Tables:');
    for (const table of expectedTables) {
      if (tables.includes(table)) {
        console.log(`  ${colors.green}✅${colors.reset} ${table}`);
      } else {
        console.log(`  ${colors.red}❌${colors.reset} ${table} (missing)`);
      }
    }

    // Check RLS status
    console.log('\nSecurity:');
    const rlsEnabled = await checkRLS();
    if (rlsEnabled) {
      console.log(`  ${colors.green}✅${colors.reset} Row Level Security enabled`);
    } else {
      console.log(`  ${colors.yellow}⚠️${colors.reset}  Row Level Security NOT enabled`);
      console.log(`     Run: ${colors.cyan}./db.sh apply${colors.reset}`);
    }

    // Show available migrations
    console.log('\nMigrations:');
    const migrations = getMigrations();
    migrations.forEach(m => {
      console.log(`  ${m.applied ? colors.green + '✅' : colors.yellow + '⏳'}${colors.reset} ${m.file}`);
    });
  },

  /**
   * Apply migrations (show SQL for manual application)
   */
  async apply() {
    console.log(`${colors.cyan}📋 Database Migrations${colors.reset}`);
    console.log('=====================\n');

    const migrations = getMigrations();

    console.log('To apply migrations, run these in Supabase Dashboard SQL Editor:\n');
    console.log('━'.repeat(60));

    for (const migration of migrations) {
      if (!migration.applied) {
        console.log(`\n${colors.yellow}-- File: ${migration.path}${colors.reset}`);
        console.log('━'.repeat(60));
        console.log(migration.content);
        console.log('━'.repeat(60));
      }
    }

    console.log(`\n${colors.cyan}After applying, run: ./db.sh status${colors.reset}`);
  },

  /**
   * Reset database (delete all data)
   */
  async reset() {
    if (!process.argv.includes('--force')) {
      console.log(`${colors.yellow}⚠️  WARNING: This will DELETE ALL DATA!${colors.reset}`);
      console.log(`Run with --force to confirm: ${colors.cyan}./db.sh reset --force${colors.reset}`);
      process.exit(1);
    }

    console.log(`${colors.red}🗑️  Resetting database...${colors.reset}\n`);

    const tables = ['audit_logs', 'webhook_events', 'usage_metrics', 'instances', 'subscriptions', 'accounts'];

    for (const table of tables) {
      try {
        const { error } = await supabase
          .from(table)
          .delete()
          .neq('id', '00000000-0000-0000-0000-000000000000');

        if (error && error.code !== '42P01') {
          console.log(`  ${colors.yellow}⚠️${colors.reset}  ${table}: ${error.message}`);
        } else {
          console.log(`  ${colors.green}✅${colors.reset} Cleared ${table}`);
        }
      } catch (e) {
        console.log(`  ${colors.red}❌${colors.reset} ${table}: ${e.message}`);
      }
    }

    console.log(`\n${colors.green}✅ Database reset complete${colors.reset}`);
    console.log(`Run ${colors.cyan}./db.sh apply${colors.reset} to reapply schema`);
  },

  /**
   * Setup Stripe products
   */
  async stripe() {
    console.log(`${colors.cyan}💳 Setting up Stripe products...${colors.reset}\n`);

    const setupScript = path.join(__dirname, 'db', 'setup-stripe-products.js');
    if (!fs.existsSync(setupScript)) {
      console.error(`${colors.red}❌ Stripe setup script not found${colors.reset}`);
      process.exit(1);
    }

    const child = spawn('node', [setupScript], { stdio: 'inherit' });
    child.on('exit', (code) => {
      if (code === 0) {
        console.log(`\n${colors.green}✅ Stripe products created${colors.reset}`);
      } else {
        console.error(`${colors.red}❌ Stripe setup failed${colors.reset}`);
      }
    });
  },

  /**
   * Create admin user
   */
  async admin() {
    console.log(`${colors.cyan}👤 Creating admin user...${colors.reset}\n`);

    const adminScript = path.join(__dirname, 'db', 'create-admin-user.js');
    if (!fs.existsSync(adminScript)) {
      console.error(`${colors.red}❌ Admin setup script not found${colors.reset}`);
      process.exit(1);
    }

    const child = spawn('node', [adminScript], { stdio: 'inherit' });
    child.on('exit', (code) => {
      if (code === 0) {
        console.log(`\n${colors.green}✅ Admin user created${colors.reset}`);
      } else {
        console.error(`${colors.red}❌ Admin setup failed${colors.reset}`);
      }
    });
  },

  /**
   * Show help
   */
  async help() {
    console.log(`${colors.cyan}🗄️  Database Manager${colors.reset}`);
    console.log('===================\n');
    console.log('Commands:');
    console.log(`  ${colors.green}status${colors.reset}  - Check database status and migrations`);
    console.log(`  ${colors.green}apply${colors.reset}   - Show SQL to apply migrations`);
    console.log(`  ${colors.green}reset${colors.reset}   - Delete all data (requires --force)`);
    console.log(`  ${colors.green}stripe${colors.reset}  - Setup Stripe products and pricing`);
    console.log(`  ${colors.green}admin${colors.reset}   - Create admin dashboard user`);
    console.log(`  ${colors.green}help${colors.reset}    - Show this help message`);
    console.log('\nUsage:');
    console.log(`  ./db.sh [command]`);
    console.log('\nExamples:');
    console.log(`  ./db.sh status`);
    console.log(`  ./db.sh reset --force`);
  }
};

// Helper functions
async function checkTables() {
  try {
    const response = await fetch(`${SUPABASE_URL}/rest/v1/`, {
      headers: {
        'apikey': SUPABASE_SERVICE_KEY,
        'Authorization': `Bearer ${SUPABASE_SERVICE_KEY}`
      }
    });
    const data = await response.json();
    return Object.keys(data.definitions || {});
  } catch (e) {
    return [];
  }
}

async function checkRLS() {
  try {
    // Try to access with anon key
    const response = await fetch(`${SUPABASE_URL}/rest/v1/accounts?limit=1`, {
      headers: {
        'apikey': SUPABASE_ANON_KEY
      }
    });
    const data = await response.json();
    // If we get an empty array, RLS might not be enabled
    // If we get an error, RLS is likely enabled
    return !Array.isArray(data) || data.message;
  } catch (e) {
    return true; // Error likely means RLS is blocking
  }
}

function getMigrations() {
  const migrationsDir = path.join(__dirname, '..', 'supabase', 'migrations');
  const files = fs.readdirSync(migrationsDir)
    .filter(f => f.endsWith('.sql'))
    .sort();

  return files.map(file => {
    const content = fs.readFileSync(path.join(migrationsDir, file), 'utf8');
    const applied = file === '001_schema.sql' ? checkTables().length > 0 : false;
    return {
      file,
      path: `supabase/migrations/${file}`,
      content: content.substring(0, 500) + (content.length > 500 ? '\n...' : ''),
      applied
    };
  });
}

// Main execution
async function main() {
  const command = process.argv[2] || 'help';

  if (commands[command]) {
    try {
      await commands[command]();
    } catch (error) {
      console.error(`${colors.red}❌ Error: ${error.message}${colors.reset}`);
      process.exit(1);
    }
  } else {
    console.error(`${colors.red}❌ Unknown command: ${command}${colors.reset}`);
    await commands.help();
    process.exit(1);
  }
}

main();

File: saas-platform/scripts/db-manager.py
#!/usr/bin/env -S uv run --script
#
# /// script
# requires-python = ">=3.12"
# dependencies = ["typer", "rich", "httpx", "python-dotenv"]
# ///

"""Unified Database Management Tool."""

import os
import subprocess
import sys
from pathlib import Path

import httpx
import typer
from dotenv import load_dotenv
from rich.console import Console
from rich.panel import Panel
from rich.syntax import Syntax
from rich.table import Table

# Load environment variables
load_dotenv()

# Initialize Rich console and Typer app
console = Console()
app = typer.Typer(help="Database management tool for MindRoom SaaS platform")

# Environment variables
SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_SERVICE_KEY = os.getenv("SUPABASE_SERVICE_KEY")
SUPABASE_ANON_KEY = os.getenv("SUPABASE_ANON_KEY")

# Check required environment variables
if not SUPABASE_URL or not SUPABASE_SERVICE_KEY:
    console.print("[red]❌ Missing required environment variables:[/red]")
    console.print("   SUPABASE_URL and SUPABASE_SERVICE_KEY must be set in .env")
    sys.exit(1)

# Expected tables
EXPECTED_TABLES = [
    "accounts",
    "subscriptions",
    "instances",
    "usage_metrics",
    "webhook_events",
    "audit_logs",
]

# Table order for reset (respects foreign key constraints)
RESET_TABLE_ORDER = [
    "audit_logs",
    "webhook_events",
    "usage_metrics",
    "instances",
    "subscriptions",
    "accounts",
]


def get_migrations_dir() -> Path:
    """Get the migrations directory path."""
    return Path(__file__).parent.parent / "supabase" / "migrations"


def check_tables() -> list[str]:
    """Check which tables exist in the database."""
    try:
        response = httpx.get(
            f"{SUPABASE_URL}/rest/v1/",
            headers={
                "apikey": SUPABASE_SERVICE_KEY,
                "Authorization": f"Bearer {SUPABASE_SERVICE_KEY}",
            },
        )
        data = response.json()
        return list(data.get("definitions", {}).keys())
    except Exception:
        return []


def check_rls() -> bool:
    """Check if Row Level Security is enabled."""
    try:
        # Try to access with anon key
        response = httpx.get(
            f"{SUPABASE_URL}/rest/v1/accounts?limit=1",
            headers={"apikey": SUPABASE_ANON_KEY},
        )
        data = response.json()
        # If we get an error or message, RLS is likely enabled
        return not isinstance(data, list) or "message" in data
    except Exception:
        return True  # Error likely means RLS is blocking


def get_migrations() -> list[dict]:
    """Get list of migration files."""
    migrations_dir = get_migrations_dir()
    if not migrations_dir.exists():
        return []

    sql_files = sorted(migrations_dir.glob("*.sql"))
    migrations = []

    for file in sql_files:
        content = file.read_text()
        # Check if this migration is applied (simplified check)
        applied = file.name == "001_schema.sql" and len(check_tables()) > 0

        migrations.append(
            {
                "file": file.name,
                "path": f"supabase/migrations/{file.name}",
                "content": content[:500] + ("..." if len(content) > 500 else ""),
                "full_content": content,
                "applied": applied,
            },
        )

    return migrations


@app.command()
def status() -> None:
    """Check database status and migrations."""
    console.print(Panel.fit("[cyan]📊 Database Status[/cyan]", border_style="cyan"))

    # Check tables
    existing_tables = check_tables()
    table = Table(title="Tables", show_header=False)
    table.add_column("Status", style="dim", width=3)
    table.add_column("Table")

    for table_name in EXPECTED_TABLES:
        if table_name in existing_tables:
            table.add_row("✅", f"[green]{table_name}[/green]")
        else:
            table.add_row("❌", f"[red]{table_name} (missing)[/red]")

    console.print(table)

    # Check RLS
    console.print("\n[bold]Security:[/bold]")
    if check_rls():
        console.print("  ✅ Row Level Security enabled")
    else:
        console.print("  [yellow]⚠️  Row Level Security NOT enabled[/yellow]")
        console.print("     Run: [cyan]./db.sh apply[/cyan]")

    # Show migrations
    console.print("\n[bold]Migrations:[/bold]")
    migrations = get_migrations()
    for m in migrations:
        status_icon = "✅" if m["applied"] else "⏳"
        status_color = "green" if m["applied"] else "yellow"
        console.print(f"  {status_icon} [{status_color}]{m['file']}[/{status_color}]")


@app.command()
def apply() -> None:
    """Show SQL to apply migrations."""
    console.print(Panel.fit("[cyan]📋 Database Migrations[/cyan]", border_style="cyan"))

    migrations = get_migrations()
    unapplied = [m for m in migrations if not m["applied"]]

    if not unapplied:
        console.print("[green]✅ All migrations are applied![/green]")
        return

    console.print("To apply migrations, run these in Supabase Dashboard SQL Editor:\n")

    for migration in unapplied:
        console.print(f"[yellow]-- File: {migration['path']}[/yellow]")
        console.print("─" * 60)

        # Use Syntax for SQL highlighting
        syntax = Syntax(migration["full_content"], "sql", theme="monokai", line_numbers=True)
        console.print(syntax)
        console.print("─" * 60)
        console.print()

    console.print("[cyan]After applying, run: ./db.sh status[/cyan]")


@app.command()
def reset(force: bool = typer.Option(False, "--force", help="Confirm database reset")) -> None:
    """Reset database (delete all data)."""
    if not force:
        console.print("[yellow]⚠️  WARNING: This will DELETE ALL DATA![/yellow]")
        console.print("Run with --force to confirm: [cyan]./db.sh reset --force[/cyan]")
        raise typer.Exit(1)

    console.print("[red]🗑️  Resetting database...[/red]\n")

    # Create Supabase client headers
    headers = {
        "apikey": SUPABASE_SERVICE_KEY,
        "Authorization": f"Bearer {SUPABASE_SERVICE_KEY}",
        "Content-Type": "application/json",
        "Prefer": "return=minimal",
    }

    for table_name in RESET_TABLE_ORDER:
        try:
            # Delete all records from table
            response = httpx.delete(
                f"{SUPABASE_URL}/rest/v1/{table_name}",
                headers=headers,
                params={"id": "neq.00000000-0000-0000-0000-000000000000"},
            )

            if response.status_code in [200, 204]:
                console.print(f"  ✅ Cleared {table_name}")
            elif response.status_code == 404:
                console.print(f"  [yellow]⚠️[/yellow]  {table_name}: Table not found")
            else:
                console.print(f"  [yellow]⚠️[/yellow]  {table_name}: {response.text}")
        except Exception as e:
            console.print(f"  [red]❌[/red] {table_name}: {e!s}")

    console.print("\n[green]✅ Database reset complete[/green]")
    console.print("Run [cyan]./db.sh apply[/cyan] to reapply schema")


@app.command()
def stripe() -> None:
    """Setup Stripe products and pricing."""
    console.print("[cyan]💳 Setting up Stripe products...[/cyan]\n")

    setup_script = Path(__file__).parent / "db" / "setup-stripe-products.js"
    if not setup_script.exists():
        console.print("[red]❌ Stripe setup script not found[/red]")
        raise typer.Exit(1)

    try:
        result = subprocess.run(
            ["node", str(setup_script)],
            check=False,
            capture_output=False,
            text=True,
        )
        if result.returncode == 0:
            console.print("\n[green]✅ Stripe products created[/green]")
        else:
            console.print("[red]❌ Stripe setup failed[/red]")
            raise typer.Exit(1)  # noqa: TRY301
    except Exception as e:
        console.print(f"[red]❌ Error running Stripe setup: {e}[/red]")
        raise typer.Exit(1)  # noqa: B904


@app.command()
def admin() -> None:
    """Create admin dashboard user."""
    console.print("[cyan]👤 Creating admin user...[/cyan]\n")

    admin_script = Path(__file__).parent / "db" / "create-admin-user.js"
    if not admin_script.exists():
        console.print("[red]❌ Admin setup script not found[/red]")
        raise typer.Exit(1)

    try:
        result = subprocess.run(
            ["node", str(admin_script)],
            check=False,
            capture_output=False,
            text=True,
        )
        if result.returncode == 0:
            console.print("\n[green]✅ Admin user created[/green]")
        else:
            console.print("[red]❌ Admin setup failed[/red]")
            raise typer.Exit(1)  # noqa: TRY301
    except Exception as e:
        console.print(f"[red]❌ Error creating admin user: {e}[/red]")
        raise typer.Exit(1)  # noqa: B904


if __name__ == "__main__":
    app()
